
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      <link rel="icon" href="../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.2.3, mkdocs-material-8.1.8">
    
    
      
        <title>Annotate a VCF with GraphKB - PORI</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.6e60f8b8.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.e6a45f82.min.css">
        
      
    
    
    
      
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../styles.css">
    
      <link rel="stylesheet" href="../../../css/ansi-colours.css">
    
      <link rel="stylesheet" href="../../../css/jupyter-cells.css">
    
      <link rel="stylesheet" href="../../../css/pandas-dataframe.css">
    
    <script>__md_scope=new URL("../../..",location),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="none" data-md-color-accent="none">
  
    
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#annotate-a-vcf-with-graphkb" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="PORI" class="md-header__button md-logo" aria-label="PORI" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            PORI
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Annotate a VCF with GraphKB
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/bcgsc/pori/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    github
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  

<nav class="md-nav md-nav--primary md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="PORI" class="md-nav__button md-logo" aria-label="PORI" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    PORI
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/bcgsc/pori/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    github
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../../demo/" class="md-nav__link">
        Demo
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../../install/" class="md-nav__link">
        Install with Docker
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../../repos/" class="md-nav__link">
        All PORI Repositories
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    
      
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_5" type="checkbox" id="__nav_5" checked>
      
      
      
      
        <label class="md-nav__link" for="__nav_5">
          GraphKB
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="GraphKB" data-md-level="1">
        <label class="md-nav__title" for="__nav_5">
          <span class="md-nav__icon md-icon"></span>
          GraphKB
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../" class="md-nav__link">
        About GraphKB
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../loading_data/" class="md-nav__link">
        Loading Data
      </a>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_5_3" type="checkbox" id="__nav_5_3" >
      
      
      
      
        <label class="md-nav__link" for="__nav_5_3">
          How Matching Works
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="How Matching Works" data-md-level="2">
        <label class="md-nav__title" for="__nav_5_3">
          <span class="md-nav__icon md-icon"></span>
          How Matching Works
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../matching/" class="md-nav__link">
        Ontology Algorithm
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../matching/gene/" class="md-nav__link">
        Gene Example
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_5_4" type="checkbox" id="__nav_5_4" checked>
      
      
      
      
        <label class="md-nav__link" for="__nav_5_4">
          Scripting Examples
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Scripting Examples" data-md-level="2">
        <label class="md-nav__title" for="__nav_5_4">
          <span class="md-nav__icon md-icon"></span>
          Scripting Examples
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../" class="md-nav__link">
        About the Python Adapter
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../Intro_Tutorial/" class="md-nav__link">
        Introductory Tutorial
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../Query_Basics/" class="md-nav__link">
        Query Basics
      </a>
    </li>
  

            
          
            
              
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Annotate a VCF with GraphKB
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Annotate a VCF with GraphKB
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#download-files" class="md-nav__link">
    Download Files
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#install" class="md-nav__link">
    Install
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#prepare-canonical-transcripts-list" class="md-nav__link">
    Prepare Canonical Transcripts List
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#process-the-vcf" class="md-nav__link">
    Process the VCF
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#annotate" class="md-nav__link">
    Annotate
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../snpsift/" class="md-nav__link">
        Annotate SNPSift Files
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../variant_strings/" class="md-nav__link">
        Annotate Variant List
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
      
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_6" type="checkbox" id="__nav_6" >
      
      
      
      
        <label class="md-nav__link" for="__nav_6">
          IPR
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="IPR" data-md-level="1">
        <label class="md-nav__title" for="__nav_6">
          <span class="md-nav__icon md-icon"></span>
          IPR
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../ipr/" class="md-nav__link">
        About IPR
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../ipr/upload/" class="md-nav__link">
        Uploading A Report
      </a>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_6_3" type="checkbox" id="__nav_6_3" >
      
      
      
      
        <label class="md-nav__link" for="__nav_6_3">
          Core Variant Types
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Core Variant Types" data-md-level="2">
        <label class="md-nav__title" for="__nav_6_3">
          <span class="md-nav__icon md-icon"></span>
          Core Variant Types
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../ipr/core_variants/" class="md-nav__link">
        About
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../ipr/core_variants/copy/" class="md-nav__link">
        Copy Variants
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../ipr/core_variants/expression/" class="md-nav__link">
        Expression Variants
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../ipr/core_variants/mutations/" class="md-nav__link">
        Small Mutations
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../ipr/core_variants/structural/" class="md-nav__link">
        Structural Variants
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../ipr/kb_matches/" class="md-nav__link">
        KB Matches
      </a>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_6_5" type="checkbox" id="__nav_6_5" >
      
      
      
      
        <label class="md-nav__link" for="__nav_6_5">
          Optional Analyses
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Optional Analyses" data-md-level="2">
        <label class="md-nav__title" for="__nav_6_5">
          <span class="md-nav__icon md-icon"></span>
          Optional Analyses
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../ipr/optional_analyses/burden/" class="md-nav__link">
        Mutation Burden
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../ipr/optional_analyses/circos_plots/" class="md-nav__link">
        Circos Plots
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../ipr/optional_analyses/comparators/" class="md-nav__link">
        Comparators
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../ipr/optional_analyses/expression_correlation/" class="md-nav__link">
        Expression Correlation
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../ipr/optional_analyses/images/" class="md-nav__link">
        Images
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../ipr/optional_analyses/immune/" class="md-nav__link">
        Immune Profiling
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../ipr/optional_analyses/metadata/" class="md-nav__link">
        Patient Metadata
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../ipr/optional_analyses/signature/" class="md-nav__link">
        Mutation Signatures
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../ipr/manual_curation/" class="md-nav__link">
        Manual Curation
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../ipr/templates/" class="md-nav__link">
        Templates
      </a>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_6_8" type="checkbox" id="__nav_6_8" >
      
      
      
      
        <label class="md-nav__link" for="__nav_6_8">
          Scripting Examples
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Scripting Examples" data-md-level="2">
        <label class="md-nav__title" for="__nav_6_8">
          <span class="md-nav__icon md-icon"></span>
          Scripting Examples
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../ipr/scripting/RNA_Expression_Metrics/" class="md-nav__link">
        RNA Expression Metrics
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../ipr/scripting/RNA_Expression_Similarity/" class="md-nav__link">
        RNA Expression Similarity
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
      
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_7" type="checkbox" id="__nav_7" >
      
      
      
      
        <label class="md-nav__link" for="__nav_7">
          Variant Notation
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Variant Notation" data-md-level="1">
        <label class="md-nav__title" for="__nav_7">
          <span class="md-nav__icon md-icon"></span>
          Variant Notation
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../variant_notation/" class="md-nav__link">
        About Variant Notation
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../variant_notation/continuous/" class="md-nav__link">
        Continuous Notation
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../variant_notation/split/" class="md-nav__link">
        Multi-Feature (Split) Notation
      </a>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_7_4" type="checkbox" id="__nav_7_4" >
      
      
      
      
        <label class="md-nav__link" for="__nav_7_4">
          Coordinate Systems
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Coordinate Systems" data-md-level="2">
        <label class="md-nav__title" for="__nav_7_4">
          <span class="md-nav__icon md-icon"></span>
          Coordinate Systems
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../variant_notation/coordinate_systems/" class="md-nav__link">
        About
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../variant_notation/coordinate_systems/cytoband/" class="md-nav__link">
        CytoBand Coordinates
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
      
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_8" type="checkbox" id="__nav_8" >
      
      
      
      
        <label class="md-nav__link" for="__nav_8">
          Help
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Help" data-md-level="1">
        <label class="md-nav__title" for="__nav_8">
          <span class="md-nav__icon md-icon"></span>
          Help
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../help/faq/" class="md-nav__link">
        General FAQ
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../help/troubleshooting/" class="md-nav__link">
        Troubleshooting
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
      
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_9" type="checkbox" id="__nav_9" >
      
      
      
      
        <label class="md-nav__link" for="__nav_9">
          Developer Reference
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Developer Reference" data-md-level="1">
        <label class="md-nav__title" for="__nav_9">
          <span class="md-nav__icon md-icon"></span>
          Developer Reference
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../developer_reference/" class="md-nav__link">
        About the Developer Reference
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../developer_reference/contributing/" class="md-nav__link">
        Contributing Guidelines
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../developer_reference/getting_started/" class="md-nav__link">
        Getting Started
      </a>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_9_4" type="checkbox" id="__nav_9_4" >
      
      
      
      
        <label class="md-nav__link" for="__nav_9_4">
          Graphkb
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Graphkb" data-md-level="2">
        <label class="md-nav__title" for="__nav_9_4">
          <span class="md-nav__icon md-icon"></span>
          Graphkb
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../developer_reference/graphkb/__init__/" class="md-nav__link">
        graphkb
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../developer_reference/graphkb/genes/" class="md-nav__link">
        graphkb.genes
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../developer_reference/graphkb/match/" class="md-nav__link">
        graphkb.match
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../developer_reference/graphkb/statement/" class="md-nav__link">
        graphkb.statement
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../developer_reference/graphkb/types/" class="md-nav__link">
        graphkb.types
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../developer_reference/graphkb/util/" class="md-nav__link">
        graphkb.util
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../developer_reference/graphkb/vocab/" class="md-nav__link">
        graphkb.vocab
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_9_5" type="checkbox" id="__nav_9_5" >
      
      
      
      
        <label class="md-nav__link" for="__nav_9_5">
          Ipr
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Ipr" data-md-level="2">
        <label class="md-nav__title" for="__nav_9_5">
          <span class="md-nav__icon md-icon"></span>
          Ipr
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../developer_reference/ipr/annotate/" class="md-nav__link">
        ipr.annotate
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../developer_reference/ipr/connection/" class="md-nav__link">
        ipr.connection
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../developer_reference/ipr/inputs/" class="md-nav__link">
        ipr.inputs
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../developer_reference/ipr/ipr/" class="md-nav__link">
        ipr.ipr
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../developer_reference/ipr/main/" class="md-nav__link">
        ipr.main
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../developer_reference/ipr/summary/" class="md-nav__link">
        ipr.summary
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../developer_reference/ipr/therapeutic_options/" class="md-nav__link">
        ipr.therapeutic_options
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../developer_reference/ipr/types/" class="md-nav__link">
        ipr.types
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../developer_reference/ipr/util/" class="md-nav__link">
        ipr.util
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                
<a href="https://github.com/bcgsc/pori/edit/master/docs/graphkb/scripting/Annotate_a_VCF_with_GraphKB.ipynb" title="Edit this page" class="md-content__button md-icon">
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg>
</a>


<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.1.10/require.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/2.0.3/jquery.min.js"></script>
<script>
(function() {
  function addWidgetsRenderer() {
    var mimeElement = document.querySelector('script[type="application/vnd.jupyter.widget-view+json"]');
    var scriptElement = document.createElement('script');
    var widgetRendererSrc = 'https://unpkg.com/@jupyter-widgets/html-manager@*/dist/embed-amd.js';
    var widgetState;

    // Fallback for older version:
    try {
      widgetState = mimeElement && JSON.parse(mimeElement.innerHTML);

      if (widgetState && (widgetState.version_major < 2 || !widgetState.version_major)) {
        widgetRendererSrc = 'jupyter-js-widgets@*/dist/embed.js';
      }
    } catch(e) {}

    scriptElement.src = widgetRendererSrc;
    document.body.appendChild(scriptElement);
  }

  document.addEventListener('DOMContentLoaded', addWidgetsRenderer);
}());
</script>

<div class="cell border-box-sizing text_cell rendered">
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="annotate-a-vcf-with-graphkb">Annotate a VCF with GraphKB<a class="headerlink" href="#annotate-a-vcf-with-graphkb" title="Permanent link">&para;</a></h1>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p><a href="https://colab.research.google.com/github/bcgsc/pori/blob/feature%2Fcolab-notebooks/docs/graphkb/scripting/Annotate_a_VCF_with_GraphKB.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a></p>
<p>A common format for small mutations is the VCF file. There are a number of variants of this format. To help users getting started we have provided this example script of how to convert from a VCF file to the JSON format that the IPR python adapter expects.</p>
<p>We expect users to annotate their VCF files with protein and/or cds HGVS notation prior to this.</p>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="download-files">Download Files<a class="headerlink" href="#download-files" title="Permanent link">&para;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="install">Install<a class="headerlink" href="#install" title="Permanent link">&para;</a></h2>
<p>To do this we are going to install a couple of common python libraries. We will use pysam to read the vcf file (<a href="https://pysam.readthedocs.io/en/latest/usage.html#working-with-vcf-bcf-formatted-files">Pysam - Working with VCF</a>).</p>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="highlight"><pre><span></span><code><span class="err">!</span><span class="n">pip</span> <span class="n">install</span> <span class="n">pysam</span> <span class="n">pandas</span> <span class="n">seaborn</span> <span class="n">graphkb</span>
</code></pre></div>

</div>
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="output_subarea output_stream output_stdout output_text">
<pre>
<code>Collecting pysam
  Downloading https://files.pythonhosted.org/packages/20/85/335857b9888f6d9a13b03a8f21b0a6228b180c361631d9d70e7be3e22163/pysam-0.16.0.1-cp37-cp37m-manylinux1_x86_64.whl (9.9MB)
     |████████████████████████████████| 9.9MB 4.3MB/s 
Requirement already satisfied: pandas in /usr/local/lib/python3.7/dist-packages (1.1.5)
Requirement already satisfied: seaborn in /usr/local/lib/python3.7/dist-packages (0.11.1)
Collecting graphkb
  Downloading https://files.pythonhosted.org/packages/cc/34/48ec589f81fce66cc70053e666a7834c3a7f708cee9a101dfc3d90158fe6/graphkb-1.5.4-py3-none-any.whl
Requirement already satisfied: pytz&gt;=2017.2 in /usr/local/lib/python3.7/dist-packages (from pandas) (2018.9)
Requirement already satisfied: python-dateutil&gt;=2.7.3 in /usr/local/lib/python3.7/dist-packages (from pandas) (2.8.1)
Requirement already satisfied: numpy&gt;=1.15.4 in /usr/local/lib/python3.7/dist-packages (from pandas) (1.19.5)
Requirement already satisfied: matplotlib&gt;=2.2 in /usr/local/lib/python3.7/dist-packages (from seaborn) (3.2.2)
Requirement already satisfied: scipy&gt;=1.0 in /usr/local/lib/python3.7/dist-packages (from seaborn) (1.4.1)
Requirement already satisfied: typing-extensions&lt;4,&gt;=3.7.4.2 in /usr/local/lib/python3.7/dist-packages (from graphkb) (3.7.4.3)
Requirement already satisfied: requests&lt;3,&gt;=2.22.0 in /usr/local/lib/python3.7/dist-packages (from graphkb) (2.23.0)
Requirement already satisfied: six&gt;=1.5 in /usr/local/lib/python3.7/dist-packages (from python-dateutil&gt;=2.7.3-&gt;pandas) (1.15.0)
Requirement already satisfied: pyparsing!=2.0.4,!=2.1.2,!=2.1.6,&gt;=2.0.1 in /usr/local/lib/python3.7/dist-packages (from matplotlib&gt;=2.2-&gt;seaborn) (2.4.7)
Requirement already satisfied: cycler&gt;=0.10 in /usr/local/lib/python3.7/dist-packages (from matplotlib&gt;=2.2-&gt;seaborn) (0.10.0)
Requirement already satisfied: kiwisolver&gt;=1.0.1 in /usr/local/lib/python3.7/dist-packages (from matplotlib&gt;=2.2-&gt;seaborn) (1.3.1)
Requirement already satisfied: chardet&lt;4,&gt;=3.0.2 in /usr/local/lib/python3.7/dist-packages (from requests&lt;3,&gt;=2.22.0-&gt;graphkb) (3.0.4)
Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,&lt;1.26,&gt;=1.21.1 in /usr/local/lib/python3.7/dist-packages (from requests&lt;3,&gt;=2.22.0-&gt;graphkb) (1.24.3)
Requirement already satisfied: idna&lt;3,&gt;=2.5 in /usr/local/lib/python3.7/dist-packages (from requests&lt;3,&gt;=2.22.0-&gt;graphkb) (2.10)
Requirement already satisfied: certifi&gt;=2017.4.17 in /usr/local/lib/python3.7/dist-packages (from requests&lt;3,&gt;=2.22.0-&gt;graphkb) (2021.5.30)
Installing collected packages: pysam, graphkb
Successfully installed graphkb-1.5.4 pysam-0.16.0.1
</code>
</pre>
</div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="prepare-canonical-transcripts-list">Prepare Canonical Transcripts List<a class="headerlink" href="#prepare-canonical-transcripts-list" title="Permanent link">&para;</a></h2>
<p>First we will read the tabbed biomart export file. This contains a list of gene and transcript IDs for only canonical transcripts. We'll use this to filter annotations in a later step</p>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="highlight"><pre><span></span><code><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span> 

<span class="n">mart_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;https://raw.githubusercontent.com/bcgsc/pori/feature/colab-notebooks/demo/mart_export.protein_coding.canonical.txt&#39;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="n">mart_df</span>
</code></pre></div>

</div>
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped="scoped">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
<thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Gene stable ID version</th>
      <th>Transcript stable ID version</th>
    </tr>
  </thead>
<tbody>
    <tr>
      <th>0</th>
      <td>ENSG00000198888.2</td>
      <td>ENST00000361390.2</td>
    </tr>
    <tr>
      <th>1</th>
      <td>ENSG00000198763.3</td>
      <td>ENST00000361453.3</td>
    </tr>
    <tr>
      <th>2</th>
      <td>ENSG00000198804.2</td>
      <td>ENST00000361624.2</td>
    </tr>
    <tr>
      <th>3</th>
      <td>ENSG00000198712.1</td>
      <td>ENST00000361739.1</td>
    </tr>
    <tr>
      <th>4</th>
      <td>ENSG00000228253.1</td>
      <td>ENST00000361851.1</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>22791</th>
      <td>ENSG00000181817.6</td>
      <td>ENST00000315732.3</td>
    </tr>
    <tr>
      <th>22792</th>
      <td>ENSG00000116885.18</td>
      <td>ENST00000235532.9</td>
    </tr>
    <tr>
      <th>22793</th>
      <td>ENSG00000116898.12</td>
      <td>ENST00000373116.6</td>
    </tr>
    <tr>
      <th>22794</th>
      <td>ENSG00000119535.18</td>
      <td>ENST00000373106.6</td>
    </tr>
    <tr>
      <th>22795</th>
      <td>ENSG00000142694.7</td>
      <td>ENST00000490466.2</td>
    </tr>
  </tbody>
</table>
<p>22796 rows × 2 columns</p>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>In this case we know that the transcripts in our VCF are unversioned so we will store the transcript names without their versions as well for quick access in later steps</p>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="highlight"><pre><span></span><code><span class="n">mart_df</span><span class="p">[</span><span class="s1">&#39;unversioned&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">mart_df</span><span class="p">[</span><span class="s1">&#39;Transcript stable ID version&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="n">canonical_transcripts</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">mart_df</span><span class="p">[</span><span class="s1">&#39;Transcript stable ID version&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span> <span class="o">+</span> <span class="n">mart_df</span><span class="p">[</span><span class="s1">&#39;unversioned&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
<span class="nb">len</span><span class="p">(</span><span class="n">canonical_transcripts</span><span class="p">)</span>
</code></pre></div>

</div>
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="output_text output_subarea output_execute_result">
<pre>
<code>45592</code>
</pre>
</div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="process-the-vcf">Process the VCF<a class="headerlink" href="#process-the-vcf" title="Permanent link">&para;</a></h2>
<p>Then we are ready to read the VCF file as input</p>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="highlight"><pre><span></span><code><span class="err">!</span><span class="n">wget</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">raw</span><span class="o">.</span><span class="n">githubusercontent</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">bcgsc</span><span class="o">/</span><span class="n">pori</span><span class="o">/</span><span class="n">feature</span><span class="o">/</span><span class="n">colab</span><span class="o">-</span><span class="n">notebooks</span><span class="o">/</span><span class="n">demo</span><span class="o">/</span><span class="n">example_snvs</span><span class="o">.</span><span class="n">variants</span><span class="o">.</span><span class="n">passed</span><span class="mf">.4.3</span><span class="o">.</span><span class="n">eff</span><span class="o">.</span><span class="n">hgvs</span><span class="o">.</span><span class="n">vcf</span>
</code></pre></div>

</div>
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="output_subarea output_stream output_stdout output_text">
<pre>
<code>--2021-06-17 20:42:11--  https://raw.githubusercontent.com/bcgsc/pori/feature/colab-notebooks/demo/example_snvs.variants.passed.4.3.eff.hgvs.vcf
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 387417 (378K) [text/plain]
Saving to: ‘example_snvs.variants.passed.4.3.eff.hgvs.vcf.1’

example_snvs.varian 100%[===================&gt;] 378.34K  --.-KB/s    in 0.04s   

2021-06-17 20:42:12 (8.94 MB/s) - ‘example_snvs.variants.passed.4.3.eff.hgvs.vcf.1’ saved [387417/387417]

</code>
</pre>
</div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="highlight"><pre><span></span><code><span class="kn">from</span> <span class="nn">pysam</span> <span class="kn">import</span> <span class="n">VariantFile</span>

<span class="n">fh</span> <span class="o">=</span> <span class="n">VariantFile</span><span class="p">(</span><span class="s1">&#39;/content/example_snvs.variants.passed.4.3.eff.hgvs.vcf&#39;</span><span class="p">)</span>
</code></pre></div>

</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Then we will iterate over the variant VCF records. From here we will need to pull out the HGVSp notation. Since this is output from SnpEff we can refer to their <a href="https://pcingola.github.io/SnpEff/se_inputoutput/#eff-field-vcf-output-files">documentation</a> for how to parse this. According to their site this is the expected format of the <code>EFF</code> field which contains the HGVS notation. This field is pipe delimited inside parentheses.</p>
<div class="highlight"><pre><span></span><code>EFF= Effect ( Effect_Impact | Functional_Class | Codon_Change | Amino_Acid_Change| Amino_Acid_Length | Gene_Name | Transcript_BioType | Gene_Coding | Transcript_ID | Exon_Rank  | Genotype_Number [ | ERRORS | WARNINGS ] )
</code></pre></div>
<p><strong>Note</strong>: EFF is a deprecated field. Newer versions use the <code>ANN</code> field. Parsing should be similar.</p>
<p>We define the following functions below to help with parsing the INFO fields in the VCF</p>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html"></div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="highlight"><pre><span></span><code><span class="kn">import</span> <span class="nn">re</span> 


<span class="k">def</span> <span class="nf">parse_eff</span><span class="p">(</span><span class="n">eff</span><span class="p">):</span>
  <span class="n">match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;^[^(]+\((.+)\)$&#39;</span><span class="p">,</span> <span class="n">eff</span><span class="p">)</span>
  <span class="k">if</span> <span class="n">match</span><span class="p">:</span>
    <span class="p">[</span><span class="n">impact</span><span class="p">,</span> <span class="n">functional_class</span><span class="p">,</span> <span class="n">codon_change</span><span class="p">,</span> <span class="n">aa_change</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">gene_name</span><span class="p">,</span> <span class="n">transcript_biotype</span><span class="p">,</span> <span class="n">gene_coding</span><span class="p">,</span> <span class="n">transcript_id</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">genotype_allele</span> <span class="p">]</span> <span class="o">=</span> <span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;|&#39;</span><span class="p">)[:</span><span class="mi">11</span><span class="p">]</span>

    <span class="n">changes</span> <span class="o">=</span> <span class="n">aa_change</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)</span>
    <span class="n">protein_changes</span> <span class="o">=</span> <span class="p">[</span><span class="n">p</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">changes</span> <span class="k">if</span> <span class="n">p</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;p.&#39;</span><span class="p">)]</span>
    <span class="n">cds_changes</span> <span class="o">=</span> <span class="p">[</span><span class="n">p</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">changes</span> <span class="k">if</span> <span class="n">p</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;c.&#39;</span><span class="p">)]</span>

    <span class="k">return</span> <span class="p">[</span>
      <span class="n">impact</span><span class="p">,</span> 
      <span class="n">gene_name</span><span class="p">,</span>
      <span class="n">transcript_biotype</span><span class="p">,</span>
      <span class="n">gene_coding</span><span class="p">,</span> 
      <span class="n">transcript_id</span><span class="p">,</span> 
      <span class="n">protein_changes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">if</span> <span class="n">protein_changes</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span> 
      <span class="n">cds_changes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">if</span> <span class="n">cds_changes</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
      <span class="n">genotype_allele</span>
    <span class="p">]</span>
  <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s1">&#39;cannot parse eff string. Does not match the expected pattern&#39;</span><span class="p">)</span>
</code></pre></div>

</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Now we want to convert the parsed content into a JSON/Dict to match the fields that we expect in IPR. If you are not uploading to IPR you can leave variants in their original format instead.</p>
<p><strong>Note</strong>: Some of the fields (support counts) require picking the samples. By default we have assumed the names are NORMAL and TUMOR. These may need to be customized for other input</p>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="highlight"><pre><span></span><code><span class="kn">from</span> <span class="nn">graphkb.util</span> <span class="kn">import</span> <span class="n">convert_aa_3to1</span>


<span class="k">def</span> <span class="nf">pick_effect</span><span class="p">(</span><span class="n">record</span><span class="p">):</span>
  <span class="n">non_canonical</span> <span class="o">=</span> <span class="p">[]</span>

  <span class="k">for</span> <span class="n">eff</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s1">&#39;EFF&#39;</span><span class="p">]:</span>
    <span class="p">[</span><span class="n">impact</span><span class="p">,</span> <span class="n">gene_name</span><span class="p">,</span> <span class="n">transcript_biotype</span><span class="p">,</span> <span class="n">gene_coding</span><span class="p">,</span> <span class="n">transcript_id</span><span class="p">,</span> <span class="n">protein_change</span><span class="p">,</span> <span class="n">cds_change</span><span class="p">,</span> <span class="n">alt_allele</span><span class="p">]</span> <span class="o">=</span> <span class="n">parse_eff</span><span class="p">(</span><span class="n">eff</span><span class="p">)</span>

    <span class="c1"># prefer canonical annotation</span>
    <span class="k">if</span> <span class="n">transcript_id</span> <span class="ow">in</span> <span class="n">canonical_transcripts</span><span class="p">:</span>
      <span class="k">return</span> <span class="p">[</span><span class="n">impact</span><span class="p">,</span> <span class="n">gene_name</span><span class="p">,</span> <span class="n">transcript_biotype</span><span class="p">,</span> <span class="n">gene_coding</span><span class="p">,</span> <span class="n">transcript_id</span><span class="p">,</span> <span class="n">protein_change</span><span class="p">,</span> <span class="n">cds_change</span><span class="p">,</span> <span class="n">alt_allele</span><span class="p">]</span>
    <span class="n">non_canonical</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">impact</span><span class="p">,</span> <span class="n">gene_name</span><span class="p">,</span> <span class="n">transcript_biotype</span><span class="p">,</span> <span class="n">gene_coding</span><span class="p">,</span> <span class="n">transcript_id</span><span class="p">,</span> <span class="n">protein_change</span><span class="p">,</span> <span class="n">cds_change</span><span class="p">,</span> <span class="n">alt_allele</span><span class="p">])</span>

  <span class="k">return</span> <span class="n">non_canonical</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>



<span class="n">ipr_variants</span> <span class="o">=</span> <span class="p">[]</span>

<span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">fh</span><span class="o">.</span><span class="n">fetch</span><span class="p">():</span>
  <span class="n">tumour_sample</span> <span class="o">=</span> <span class="n">get_sample_by_name</span><span class="p">(</span><span class="n">record</span><span class="p">,</span> <span class="s1">&#39;TUMOR&#39;</span><span class="p">)</span>
  <span class="n">normal_sample</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">samples</span><span class="p">[</span><span class="s1">&#39;NORMAL&#39;</span><span class="p">]</span>
  <span class="n">ref_key</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">record</span><span class="o">.</span><span class="n">ref</span><span class="si">}</span><span class="s1">U&#39;</span>

  <span class="n">no_canonical</span> <span class="o">=</span> <span class="kc">True</span>

  <span class="p">[</span><span class="n">impact</span><span class="p">,</span> <span class="n">gene_name</span><span class="p">,</span> <span class="n">transcript_biotype</span><span class="p">,</span> <span class="n">gene_coding</span><span class="p">,</span> <span class="n">transcript_id</span><span class="p">,</span> <span class="n">protein_change</span><span class="p">,</span> <span class="n">cds_change</span><span class="p">,</span> <span class="n">alt_allele</span><span class="p">]</span> <span class="o">=</span> <span class="n">pick_effect</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>



  <span class="n">alt_key</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">alt_allele</span><span class="si">}</span><span class="s1">U&#39;</span>
  <span class="k">if</span> <span class="n">protein_change</span><span class="p">:</span>
    <span class="n">protein_change</span> <span class="o">=</span> <span class="n">convert_aa_3to1</span><span class="p">(</span><span class="n">protein_change</span><span class="p">)</span>


  <span class="c1"># convert variant string from 3 to 1 letter AA if it is protein notation</span>
  <span class="n">ipr_format</span> <span class="o">=</span> <span class="p">{</span>
      <span class="s1">&#39;gene&#39;</span><span class="p">:</span> <span class="n">gene_name</span><span class="p">,</span>
      <span class="s1">&#39;transcript&#39;</span><span class="p">:</span> <span class="n">transcript_id</span><span class="p">,</span>
      <span class="s1">&#39;chromosome&#39;</span><span class="p">:</span> <span class="n">record</span><span class="o">.</span><span class="n">chrom</span><span class="p">,</span>
      <span class="s1">&#39;startPosition&#39;</span><span class="p">:</span> <span class="n">record</span><span class="o">.</span><span class="n">pos</span><span class="p">,</span>
      <span class="s1">&#39;endPosition&#39;</span><span class="p">:</span> <span class="n">record</span><span class="o">.</span><span class="n">pos</span> <span class="o">+</span> <span class="n">record</span><span class="o">.</span><span class="n">rlen</span><span class="p">,</span>
      <span class="s1">&#39;proteinChange&#39;</span><span class="p">:</span> <span class="n">protein_change</span> <span class="ow">or</span> <span class="n">cds_change</span><span class="p">,</span>  <span class="c1"># if no protein change use cds here instead</span>
      <span class="s1">&#39;refSeq&#39;</span><span class="p">:</span> <span class="n">record</span><span class="o">.</span><span class="n">ref</span><span class="p">,</span>
      <span class="s1">&#39;altSeq&#39;</span><span class="p">:</span> <span class="n">alt_allele</span><span class="p">,</span>
      <span class="s1">&#39;tumourDepth&#39;</span><span class="p">:</span> <span class="n">tumour_sample</span><span class="p">[</span><span class="s1">&#39;DP&#39;</span><span class="p">],</span>
      <span class="s1">&#39;normalDepth&#39;</span><span class="p">:</span> <span class="n">normal_sample</span><span class="p">[</span><span class="s1">&#39;DP&#39;</span><span class="p">],</span>
      <span class="s1">&#39;normalRefCount&#39;</span><span class="p">:</span> <span class="n">normal_sample</span><span class="p">[</span><span class="n">ref_key</span><span class="p">],</span>
      <span class="s1">&#39;normalAltCount&#39;</span><span class="p">:</span> <span class="n">normal_sample</span><span class="p">[</span><span class="n">alt_key</span><span class="p">],</span>
      <span class="s1">&#39;tumourRefCount&#39;</span><span class="p">:</span> <span class="n">tumour_sample</span><span class="p">[</span><span class="n">ref_key</span><span class="p">],</span>
      <span class="s1">&#39;tumourAltCount&#39;</span><span class="p">:</span> <span class="n">tumour_sample</span><span class="p">[</span><span class="n">alt_key</span><span class="p">],</span>
      <span class="s1">&#39;hgvsProtein&#39;</span><span class="p">:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">gene_name</span><span class="si">}</span><span class="s1">:</span><span class="si">{</span><span class="n">protein_change</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">if</span> <span class="n">protein_change</span> <span class="k">else</span> <span class="s1">&#39;&#39;</span><span class="p">,</span>
      <span class="s1">&#39;hgvsCds&#39;</span><span class="p">:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">transcript_id</span><span class="si">}</span><span class="s1">:</span><span class="si">{</span><span class="n">cds_change</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">if</span> <span class="n">cds_change</span> <span class="k">else</span> <span class="s1">&#39;&#39;</span>
  <span class="p">}</span>
  <span class="n">ipr_variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ipr_format</span><span class="p">)</span>

<span class="c1"># just for displaying in notebook</span>
<span class="n">variants_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span><span class="n">ipr_variants</span><span class="p">)</span>
<span class="n">variants_df</span>
</code></pre></div>

</div>
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped="scoped">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
<thead>
    <tr style="text-align: right;">
      <th></th>
      <th>gene</th>
      <th>transcript</th>
      <th>chromosome</th>
      <th>startPosition</th>
      <th>endPosition</th>
      <th>proteinChange</th>
      <th>refSeq</th>
      <th>altSeq</th>
      <th>tumourDepth</th>
      <th>normalDepth</th>
      <th>normalRefCount</th>
      <th>normalAltCount</th>
      <th>tumourRefCount</th>
      <th>tumourAltCount</th>
      <th>hgvsProtein</th>
      <th>hgvsCds</th>
    </tr>
  </thead>
<tbody>
    <tr>
      <th>0</th>
      <td>NOC2L</td>
      <td>ENST00000327044</td>
      <td>1</td>
      <td>883516</td>
      <td>883517</td>
      <td>p.L552L</td>
      <td>G</td>
      <td>A</td>
      <td>69</td>
      <td>44</td>
      <td>(42, 45)</td>
      <td>(0, 0)</td>
      <td>(16, 16)</td>
      <td>(53, 56)</td>
      <td>NOC2L:p.L552L</td>
      <td>ENST00000327044:c.1654C&gt;T</td>
    </tr>
    <tr>
      <th>1</th>
      <td>ATAD3B</td>
      <td>ENST00000378741</td>
      <td>1</td>
      <td>1418472</td>
      <td>1418473</td>
      <td>p.R114*</td>
      <td>A</td>
      <td>T</td>
      <td>41</td>
      <td>33</td>
      <td>(31, 33)</td>
      <td>(0, 0)</td>
      <td>(29, 34)</td>
      <td>(10, 14)</td>
      <td>ATAD3B:p.R114*</td>
      <td>ENST00000378741:c.340A&gt;T</td>
    </tr>
    <tr>
      <th>2</th>
      <td>PRDM16</td>
      <td>ENST00000270722</td>
      <td>1</td>
      <td>3329312</td>
      <td>3329313</td>
      <td>p.P851S</td>
      <td>C</td>
      <td>T</td>
      <td>52</td>
      <td>49</td>
      <td>(47, 49)</td>
      <td>(2, 2)</td>
      <td>(40, 44)</td>
      <td>(10, 10)</td>
      <td>PRDM16:p.P851S</td>
      <td>ENST00000270722:c.2551C&gt;T</td>
    </tr>
    <tr>
      <th>3</th>
      <td>TNFRSF8</td>
      <td>ENST00000263932</td>
      <td>1</td>
      <td>12170228</td>
      <td>12170229</td>
      <td>p.P215S</td>
      <td>C</td>
      <td>T</td>
      <td>57</td>
      <td>26</td>
      <td>(24, 27)</td>
      <td>(0, 1)</td>
      <td>(12, 12)</td>
      <td>(43, 44)</td>
      <td>TNFRSF8:p.P215S</td>
      <td>ENST00000263932:c.643C&gt;T</td>
    </tr>
    <tr>
      <th>4</th>
      <td>CLCNKB</td>
      <td>ENST00000375679</td>
      <td>1</td>
      <td>16372084</td>
      <td>16372085</td>
      <td>p.F44F</td>
      <td>C</td>
      <td>T</td>
      <td>70</td>
      <td>51</td>
      <td>(51, 52)</td>
      <td>(0, 0)</td>
      <td>(50, 51)</td>
      <td>(18, 19)</td>
      <td>CLCNKB:p.F44F</td>
      <td>ENST00000375679:c.132C&gt;T</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>410</th>
      <td>ZNF280C</td>
      <td>ENST00000370978</td>
      <td>X</td>
      <td>129362963</td>
      <td>129362964</td>
      <td>p.H379Y</td>
      <td>G</td>
      <td>A</td>
      <td>40</td>
      <td>19</td>
      <td>(19, 21)</td>
      <td>(0, 0)</td>
      <td>(0, 0)</td>
      <td>(39, 42)</td>
      <td>ZNF280C:p.H379Y</td>
      <td>ENST00000370978:c.1135C&gt;T</td>
    </tr>
    <tr>
      <th>411</th>
      <td>ZNF280C</td>
      <td>ENST00000370978</td>
      <td>X</td>
      <td>129362964</td>
      <td>129362965</td>
      <td>p.P378P</td>
      <td>G</td>
      <td>A</td>
      <td>40</td>
      <td>19</td>
      <td>(19, 21)</td>
      <td>(0, 0)</td>
      <td>(0, 1)</td>
      <td>(39, 41)</td>
      <td>ZNF280C:p.P378P</td>
      <td>ENST00000370978:c.1134C&gt;T</td>
    </tr>
    <tr>
      <th>412</th>
      <td>PNMA3</td>
      <td>ENST00000370264</td>
      <td>X</td>
      <td>152226437</td>
      <td>152226438</td>
      <td>p.P342L</td>
      <td>C</td>
      <td>T</td>
      <td>60</td>
      <td>28</td>
      <td>(28, 32)</td>
      <td>(0, 0)</td>
      <td>(0, 1)</td>
      <td>(57, 61)</td>
      <td>PNMA3:p.P342L</td>
      <td>ENST00000370264:c.1025C&gt;T</td>
    </tr>
    <tr>
      <th>413</th>
      <td>WASH6P</td>
      <td>ENST00000359512</td>
      <td>X</td>
      <td>155253820</td>
      <td>155253821</td>
      <td>p.G373S</td>
      <td>G</td>
      <td>A</td>
      <td>74</td>
      <td>40</td>
      <td>(39, 59)</td>
      <td>(0, 1)</td>
      <td>(61, 114)</td>
      <td>(9, 20)</td>
      <td>WASH6P:p.G373S</td>
      <td>ENST00000359512:c.1117G&gt;A</td>
    </tr>
    <tr>
      <th>414</th>
      <td>AC134882.1</td>
      <td>ENST00000331172</td>
      <td>Y</td>
      <td>13524548</td>
      <td>13524549</td>
      <td>p.Y57S</td>
      <td>T</td>
      <td>G</td>
      <td>15</td>
      <td>15</td>
      <td>(13, 13)</td>
      <td>(2, 3)</td>
      <td>(0, 0)</td>
      <td>(14, 26)</td>
      <td>AC134882.1:p.Y57S</td>
      <td>ENST00000331172:c.170A&gt;C</td>
    </tr>
  </tbody>
</table>
<p>415 rows × 16 columns</p>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="annotate">Annotate<a class="headerlink" href="#annotate" title="Permanent link">&para;</a></h2>
<p>Now we are ready to add GraphKB annotations. First connect to the API. Here we are using the demo credentials. </p>
<p>If you were uploading to IPR you would skip this step and just input the variants. The IPR python adapter performs matching against GraphKB prior to uploading the report.</p>
<p><strong>NOTE</strong>: The demo server is incomplete. It contains a small subset of the data we would expect in a normal production instance of GraphKB and should be used for demonstration and testing only</p>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="highlight"><pre><span></span><code><span class="kn">from</span> <span class="nn">graphkb</span> <span class="kn">import</span> <span class="n">GraphKBConnection</span>

<span class="n">GKB_API_URL</span> <span class="o">=</span> <span class="s1">&#39;https://pori-demo.bcgsc.ca/graphkb-api/api&#39;</span>
<span class="n">GKB_USER</span> <span class="o">=</span> <span class="s1">&#39;colab_demo&#39;</span>
<span class="n">GKB_PASSWORD</span> <span class="o">=</span> <span class="s1">&#39;colab_demo&#39;</span>

<span class="n">graphkb_conn</span> <span class="o">=</span> <span class="n">GraphKBConnection</span><span class="p">(</span><span class="n">GKB_API_URL</span><span class="p">,</span> <span class="n">use_global_cache</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="n">graphkb_conn</span><span class="o">.</span><span class="n">login</span><span class="p">(</span><span class="n">GKB_USER</span><span class="p">,</span> <span class="n">GKB_PASSWORD</span><span class="p">)</span>
</code></pre></div>

</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Now use the matching functions to get the equivalent variant forms from GraphKB. This may take several minutes on the demo server as it has limited resources and open access.</p>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="highlight"><pre><span></span><code><span class="kn">from</span> <span class="nn">graphkb.match</span> <span class="kn">import</span> <span class="n">match_positional_variant</span>
<span class="kn">from</span> <span class="nn">graphkb.util</span> <span class="kn">import</span> <span class="n">FeatureNotFoundError</span>

<span class="k">for</span> <span class="n">variant</span> <span class="ow">in</span> <span class="n">ipr_variants</span><span class="p">:</span>
  <span class="c1"># for the purposes of this tutorial we will skip non-protein changes for now</span>
  <span class="k">if</span> <span class="ow">not</span> <span class="n">variant</span><span class="p">[</span><span class="s1">&#39;hgvsProtein&#39;</span><span class="p">]:</span>
    <span class="k">continue</span>
  <span class="n">variant_name</span> <span class="o">=</span> <span class="n">variant</span><span class="p">[</span><span class="s1">&#39;gene&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;:&#39;</span> <span class="o">+</span> <span class="n">variant</span><span class="p">[</span><span class="s1">&#39;proteinChange&#39;</span><span class="p">]</span>
  <span class="k">try</span><span class="p">:</span>
    <span class="n">variant_matches</span> <span class="o">=</span> <span class="n">match_positional_variant</span><span class="p">(</span><span class="n">graphkb_conn</span><span class="p">,</span> <span class="n">variant_name</span><span class="p">)</span>
    <span class="n">variant</span><span class="p">[</span><span class="s1">&#39;_variantMatches&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">variant_matches</span>
    <span class="k">if</span> <span class="n">variant_matches</span><span class="p">:</span>
      <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">variant_name</span><span class="si">}</span><span class="s1"> matched </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">variant_matches</span><span class="p">)</span><span class="si">}</span><span class="s1"> other variant representations&#39;</span><span class="p">)</span>
  <span class="k">except</span> <span class="n">FeatureNotFoundError</span><span class="p">:</span>
    <span class="k">pass</span> <span class="c1"># if the gene isn&#39;t in the db, there will not be annotations</span>
  <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">err</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">variant_name</span><span class="p">,</span> <span class="n">err</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;queried </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">ipr_variants</span><span class="p">)</span><span class="si">}</span><span class="s1"> variants&#39;</span><span class="p">)</span>
</code></pre></div>

</div>
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="output_subarea output_stream output_stdout output_text">
<pre>
<code>SF3B1:p.P718L matched 1 other variant representations
STK19:p.D89N matched 2 other variant representations
BRAF:p.V600E matched 6 other variant representations
queried 415 variants
</code>
</pre>
</div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Now that we have matched the variant we will fetch the related statements to annotate this variant with its possible relevance</p>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="highlight"><pre><span></span><code><span class="kn">from</span> <span class="nn">graphkb.constants</span> <span class="kn">import</span> <span class="n">BASE_RETURN_PROPERTIES</span><span class="p">,</span> <span class="n">GENERIC_RETURN_PROPERTIES</span>
<span class="kn">from</span> <span class="nn">graphkb.util</span> <span class="kn">import</span> <span class="n">convert_to_rid_list</span>

<span class="c1"># return properties should be customized to the users needs</span>
<span class="n">return_props</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">BASE_RETURN_PROPERTIES</span>
    <span class="o">+</span> <span class="p">[</span><span class="s1">&#39;sourceId&#39;</span><span class="p">,</span> <span class="s1">&#39;source.name&#39;</span><span class="p">,</span> <span class="s1">&#39;source.displayName&#39;</span><span class="p">]</span>
    <span class="o">+</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;conditions.</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">GENERIC_RETURN_PROPERTIES</span><span class="p">]</span>
    <span class="o">+</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;subject.</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">GENERIC_RETURN_PROPERTIES</span><span class="p">]</span>
    <span class="o">+</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;evidence.</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">GENERIC_RETURN_PROPERTIES</span><span class="p">]</span>
    <span class="o">+</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;relevance.</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">GENERIC_RETURN_PROPERTIES</span><span class="p">]</span>
    <span class="o">+</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;evidenceLevel.</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">GENERIC_RETURN_PROPERTIES</span><span class="p">]</span>
<span class="p">)</span>

<span class="n">results</span> <span class="o">=</span> <span class="p">[]</span>

<span class="k">for</span> <span class="n">variant</span> <span class="ow">in</span> <span class="n">ipr_variants</span><span class="p">:</span>
  <span class="k">if</span> <span class="ow">not</span> <span class="n">variant</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;_variantMatches&#39;</span><span class="p">):</span>
    <span class="k">continue</span>
  <span class="n">variant_name</span> <span class="o">=</span> <span class="n">variant</span><span class="p">[</span><span class="s1">&#39;gene&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;:&#39;</span> <span class="o">+</span> <span class="n">variant</span><span class="p">[</span><span class="s1">&#39;proteinChange&#39;</span><span class="p">]</span>
  <span class="n">variant_matches</span> <span class="o">=</span> <span class="n">variant</span><span class="p">[</span><span class="s1">&#39;_variantMatches&#39;</span><span class="p">]</span>
  <span class="n">statements</span> <span class="o">=</span> <span class="n">graphkb_conn</span><span class="o">.</span><span class="n">query</span><span class="p">(</span>
      <span class="p">{</span>
          <span class="s1">&#39;target&#39;</span><span class="p">:</span> <span class="s1">&#39;Statement&#39;</span><span class="p">,</span>
          <span class="s1">&#39;filters&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;conditions&#39;</span><span class="p">:</span> <span class="n">convert_to_rid_list</span><span class="p">(</span><span class="n">variant_matches</span><span class="p">),</span> <span class="s1">&#39;operator&#39;</span><span class="p">:</span> <span class="s1">&#39;CONTAINSANY&#39;</span><span class="p">},</span>
          <span class="s1">&#39;returnProperties&#39;</span><span class="p">:</span> <span class="n">return_props</span><span class="p">,</span>
      <span class="p">}</span>
  <span class="p">)</span>
  <span class="nb">print</span><span class="p">(</span><span class="n">variant_name</span><span class="p">)</span>
  <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;annotated </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">variant_matches</span><span class="p">)</span><span class="si">}</span><span class="s1"> variant matches with </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">statements</span><span class="p">)</span><span class="si">}</span><span class="s1"> statements&#39;</span><span class="p">)</span>
  <span class="nb">print</span><span class="p">()</span>

  <span class="k">for</span> <span class="n">statement</span> <span class="ow">in</span> <span class="n">statements</span><span class="p">:</span>
      <span class="n">results</span><span class="o">.</span><span class="n">append</span><span class="p">((</span>
          <span class="n">variant_name</span><span class="p">,</span> 
          <span class="nb">len</span><span class="p">(</span><span class="n">variant_matches</span><span class="p">),</span> 
          <span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">c</span><span class="p">[</span><span class="s1">&#39;displayName&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">statement</span><span class="p">[</span><span class="s1">&#39;conditions&#39;</span><span class="p">]</span> <span class="k">if</span> <span class="n">c</span><span class="p">[</span><span class="s1">&#39;@class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;Variant&#39;</span><span class="p">)]),</span>
          <span class="n">statement</span><span class="p">[</span><span class="s1">&#39;relevance&#39;</span><span class="p">][</span><span class="s1">&#39;displayName&#39;</span><span class="p">],</span>
          <span class="n">statement</span><span class="p">[</span><span class="s1">&#39;subject&#39;</span><span class="p">][</span><span class="s1">&#39;displayName&#39;</span><span class="p">],</span>
          <span class="n">statement</span><span class="p">[</span><span class="s1">&#39;source&#39;</span><span class="p">][</span><span class="s1">&#39;displayName&#39;</span><span class="p">]</span> <span class="k">if</span> <span class="n">statement</span><span class="p">[</span><span class="s1">&#39;source&#39;</span><span class="p">]</span> <span class="k">else</span> <span class="s1">&#39;&#39;</span><span class="p">,</span>
          <span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">c</span><span class="p">[</span><span class="s1">&#39;displayName&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">statement</span><span class="p">[</span><span class="s1">&#39;evidence&#39;</span><span class="p">]])</span>
      <span class="p">))</span>
</code></pre></div>

</div>
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="output_subarea output_stream output_stdout output_text">
<pre>
<code>SF3B1:p.P718L
annotated 1 variant matches with 5 statements

STK19:p.D89N
annotated 2 variant matches with 2 statements

BRAF:p.V600E
annotated 6 variant matches with 135 statements

</code>
</pre>
</div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Now we will put this into a dataframe to display nicely in this notebook</p>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="highlight"><pre><span></span><code><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">results</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;variant_name&#39;</span><span class="p">,</span> <span class="s1">&#39;variant_matches&#39;</span><span class="p">,</span> <span class="s1">&#39;statement_variants&#39;</span><span class="p">,</span> <span class="s1">&#39;relevance&#39;</span><span class="p">,</span> <span class="s1">&#39;subject&#39;</span><span class="p">,</span> <span class="s1">&#39;statement_source&#39;</span><span class="p">,</span> <span class="s1">&#39;evidence&#39;</span><span class="p">])</span>
<span class="n">df</span>
</code></pre></div>

</div>
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped="scoped">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
<thead>
    <tr style="text-align: right;">
      <th></th>
      <th>variant_name</th>
      <th>variant_matches</th>
      <th>statement_variants</th>
      <th>relevance</th>
      <th>subject</th>
      <th>statement_source</th>
      <th>evidence</th>
    </tr>
  </thead>
<tbody>
    <tr>
      <th>0</th>
      <td>SF3B1:p.P718L</td>
      <td>1</td>
      <td>SF3B1 mutation</td>
      <td>favourable prognosis</td>
      <td>patient</td>
      <td>CIViC</td>
      <td>pmid:21995386</td>
    </tr>
    <tr>
      <th>1</th>
      <td>SF3B1:p.P718L</td>
      <td>1</td>
      <td>SF3B1 mutation</td>
      <td>unfavourable prognosis</td>
      <td>patient</td>
      <td>CIViC</td>
      <td>pmid:26837699</td>
    </tr>
    <tr>
      <th>2</th>
      <td>SF3B1:p.P718L</td>
      <td>1</td>
      <td>SF3B1 mutation</td>
      <td>unfavourable prognosis</td>
      <td>patient</td>
      <td>CIViC</td>
      <td>pmid:24943832</td>
    </tr>
    <tr>
      <th>3</th>
      <td>SF3B1:p.P718L</td>
      <td>1</td>
      <td>SF3B1 mutation</td>
      <td>unfavourable prognosis</td>
      <td>patient</td>
      <td>CIViC</td>
      <td>pmid:23086750</td>
    </tr>
    <tr>
      <th>4</th>
      <td>SF3B1:p.P718L</td>
      <td>1</td>
      <td>SF3B1 mutation</td>
      <td>unfavourable prognosis</td>
      <td>patient</td>
      <td>CIViC</td>
      <td>pmid:23568491</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>137</th>
      <td>BRAF:p.V600E</td>
      <td>6</td>
      <td>BRAF mutation</td>
      <td>resistance</td>
      <td>Irinotecan [c62040]</td>
      <td>CIViC</td>
      <td>pmid:19603024</td>
    </tr>
    <tr>
      <th>138</th>
      <td>BRAF:p.V600E</td>
      <td>6</td>
      <td>BRAF mutation</td>
      <td>resistance</td>
      <td>Oxaliplatin [c1181]</td>
      <td>CIViC</td>
      <td>pmid:19603024</td>
    </tr>
    <tr>
      <th>139</th>
      <td>BRAF:p.V600E</td>
      <td>6</td>
      <td>BRAF mutation</td>
      <td>unfavourable prognosis</td>
      <td>patient</td>
      <td>CIViC</td>
      <td>pmid:19603024</td>
    </tr>
    <tr>
      <th>140</th>
      <td>BRAF:p.V600E</td>
      <td>6</td>
      <td>BRAF mutation</td>
      <td>resistance</td>
      <td>cetuximab + chemotherapy</td>
      <td>CIViC</td>
      <td>pmid:20619739</td>
    </tr>
    <tr>
      <th>141</th>
      <td>BRAF:p.V600E</td>
      <td>6</td>
      <td>BRAF mutation</td>
      <td>resistance</td>
      <td>Cetuximab [c1723]</td>
      <td>CIViC</td>
      <td>pmid:22586653</td>
    </tr>
  </tbody>
</table>
<p>142 rows × 7 columns</p>
</div>
</div>
</div>
</div>
</div>
</div>

              
            </article>
          </div>
        </div>
        
          <a href="#" class="md-top md-icon" data-md-component="top" data-md-state="hidden">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"/></svg>
            Back to top
          </a>
        
      </main>
      
        <footer class="md-footer">
  
    <nav class="md-footer__inner md-grid" aria-label="Footer">
      
        
        <a href="../Query_Basics/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Query Basics" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Query Basics
            </div>
          </div>
        </a>
      
      
        
        <a href="../snpsift/" class="md-footer__link md-footer__link--next" aria-label="Next: Annotate SNPSift Files" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              Annotate SNPSift Files
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": "../../..", "features": ["navigation.top", "navigation.sections", "navigation.instant", "toc.integrate"], "translations": {"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing", "select.version.title": "Select version"}, "search": "../../../assets/javascripts/workers/search.22074ed6.min.js"}</script>
    
    
      <script src="../../../assets/javascripts/bundle.960e086b.min.js"></script>
      
        <script src="../../../js/mathjax.config.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  </body>
</html>